crates/composition/src/emit_federated_graph.rs

```diff
@@ -129,3 +144,70 @@ fn emit_union_members(
         }
     }
 }
+
+fn emit_keys(keys: &[KeyIr], ctx: &mut Context<'_>) {
+    for KeyIr { object_id, key_id } in keys {
+        let parent_id = federated::Definition::Object(*object_id);
+        let key = ctx.subgraphs.walk(*key_id);
+        let selection = attach_selection(key.fields(), parent_id, ctx);
+        ctx.out[*object_id].resolvable_keys.push(federated::Key {
+            subgraph_id: federated::SubgraphId(key.parent_definition().subgraph().id.idx()),
+            fields: selection,
+        });
+    }
+}
+
+/// Attach a selection set defined in strings to a FederatedGraph, transforming the strings into
+/// field ids.
+fn attach_selection(
+    selection_set: &[subgraphs::Selection],
+    parent_id: federated::Definition,
+    ctx: &mut Context<'_>,
+) -> federated::SelectionSet {
+    selection_set
+        .iter()
+        .map(|selection| {
+            let field = match parent_id {
+                federated::Definition::Object(object_id) => ctx
+                    .out
+                    .object_fields
+                    .iter()
+                    .find(|object_field| {
+                        let field_name = ctx.out[object_field.field_id].name;
+                        object_field.object_id == object_id
+                            && ctx.strings_ir[field_name]
```

Comment:
I don't like that you can index by both ctx.strings_ir and ctx.out here, and only the first one is correct. I'll fix that now or later depending on feedback. There's an identical situation with FieldTypesMap.

Reply:
(because ctx.out is the federated graph and we haven't put the strings in there yet, in these functions, they're still in strings_ir at that point in time)

Reply:
Fixed in 083e6f3dcd370e8d01cde4f120f26ef83e34b0c1 — it does feel much better now. Strings and field types are in a single place.